### PR TITLE
Add capability to parse and check user-given bases mask (demux_reads …

### PR DESCRIPTION
**Pull request by messersc**
…from web) for bcl2fastq2 and picard

### CODE REVIEW

File: digestiflow_demux/tests/test_bases_mask.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+    """demux_reads is compatible with planned reads"""
+    planned_reads = "100T8B8B100T"
+    user_mask = "100T8B7B1S100T"
+    assert compare_bases_mask(planned_reads, user_mask)
+
+
+def test_compatibility2():
+    planned_reads = "100T8B8B100T"
+    user_mask = "90T10S8B8B100T"
+    assert compare_bases_mask(planned_reads, user_mask)
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply digestiflow_demux/tests/test_bases_mask.py]: See https://github.com/fossasia/query-server/issues/332#issuecomment-357734986 for a workaround for tests.

File: digestiflow_demux/bases_mask.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    # Check that no letters appear next to each other
+    diffs = []
+    for i in range(len(splitat) - 1):
+        diffs.append(splitat[i + 1] - splitat[i])
+    if (0 in diffs) or (1 in diffs):
+        raise BaseMaskConfigException("Type characters must be separated by a number (of cycles)")
+
+    result = []
+    num = ""
+    for i in range(len(bases_mask)):
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply digestiflow_demux/bases_mask.py]: agreed

File: digestiflow_demux/bases_mask.py [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    planned = split_bases_mask(planned_reads)
+    mask = split_bases_mask(bases_mask)
+
+    lengths1 = [i[1] for i in planned]
+    lengths2 = [i[1] for i in mask]
+    if not sum(lengths1) == sum(lengths2):
+        raise BaseMaskConfigException("Your base mask has more or fewer cycles than planned")
+
+    matched_mask = []
+    for type, cycles in planned:
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply digestiflow_demux/bases_mask.py]: `type` will shadow the Python buildin also, Pythonic way is `type_`.

File: digestiflow_demux/tests/test_bases_mask.py [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+def test_parse1():
+    bases_mask1 = "100T8B8B100T"
+    x = split_bases_mask(bases_mask1)
+    assert [("T", 100), ("B", 8), ("B", 8), ("T", 100)] == x
+
+
+def test_parse2():
+    bases_mask1 = "100T8B7B1S100T"
+    x = split_bases_mask(bases_mask1)
+    assert [("T", 100), ("B", 8), ("B", 7), ("S", 1), ("T", 100)] == x
```
[the remark this answers is not in the document]
Comment (holtgrewe) [reply digestiflow_demux/tests/test_bases_mask.py]: Could the split function also return a tuple?

**Comment by holtgrewe** [comment]
Generally looks good, we need a tweak for the CI and minor changes needed to code.

File: digestiflow_demux/bases_mask.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+    """Raise if base mask for demux_reads is malformed"""
+
+    pass
+
+
+def split_bases_mask(bases_mask):
+    """Parse a picard style bases mask
+    and return a list of tuples
+    e.g. for paired-end dual-indexed runs
+    [('T', 100), ('B', 8), ('B', 8), ('T', 100)]"""
```
Comment (holtgrewe) [inline digestiflow_demux/bases_mask.py]: Hm, the comment could use a bit more columns ;) Consider simply adding a doctest

https://docs.python.org/3/library/doctest.html

File: digestiflow_demux/bases_mask.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    splitat = []
+    for i, c in enumerate(bases_mask):
+        if c.isalpha():
+            splitat.append(i)
```
Comment (holtgrewe) [inline digestiflow_demux/bases_mask.py:15]: Could also be done as here https://github.com/bihealth/digestiflow-web/blob/1593d43d2535d0e2115b19cd1a20a0aea4343fa4/flowcells/models.py#L331

File: digestiflow_demux/bases_mask.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+
+def compare_bases_mask(planned_reads, bases_mask):
+    """Match user input bases mask to planned_reads from flowcell
+    and decide if compatible. Return list of lists of tuples with type and number of cycles"""
+
+    planned = split_bases_mask(planned_reads)
+    mask = split_bases_mask(bases_mask)
+
+    lengths1 = [i[1] for i in planned]
```
Comment (holtgrewe) [inline digestiflow_demux/bases_mask.py]: `[count for _type, count in planned]` would be a bit more Pythonic IMO

File: digestiflow_demux/bases_mask.py [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    and return a string to either give to bcl2fastq or picard"""
+
+    if "M" in demux_reads and demux_tool == "bcl2fastq":
+        raise BaseMaskConfigException("You cannot assign UMIs ('M') if using bcl2fastq")
+
+    mask_list = compare_bases_mask(planned_reads, demux_reads)
+
+    new_mask = []
+    for lst in mask_list:
+        substr = ""
```
Comment (holtgrewe) [inline digestiflow_demux/bases_mask.py]: Why not `[translate_tuple_to_basemask(t, demux_tool) for t in lst]`?

File: digestiflow_demux/bases_mask.py (hunk 5 above)
Comment (messersc) [reply digestiflow_demux/bases_mask.py]: Good idea. done.

File: digestiflow_demux/bases_mask.py (hunk 6 above)
Comment (messersc) [reply digestiflow_demux/bases_mask.py:15]: If you allow, I would keep it as is without using regular expressions.

**Comment by messersc** [comment]
> Generally looks good, we need a tweak for the CI.

@holtgrewe  Is there a better way than adding `noqa` to the offending `snakemake.` lines?

File: digestiflow_demux/bases_mask.py (hunk 6 above)
Comment (holtgrewe) [reply digestiflow_demux/bases_mask.py:15]: OK

**Comment by holtgrewe** [comment]
@messersc bandit should only be tweaked for tests

**Comment by holtgrewe** [comment]
@messersc `noqa` is the way to go

**Comment by messersc** [comment]
> bandit should only be tweaked for tests

As I understand it, it is only possible to either skip a class of tests or to exclude files/directories. But it is currently not possible to only disable a test (`B101`) for a given directory (`tests`)

https://github.com/PyCQA/bandit/issues/346

**Comment by holtgrewe** [comment]
OK, fine.

**Comment by holtgrewe** [comment]
Please make sure to squash and use a rebase commit.

**Pull request merged**